### PR TITLE
PM-Xpath implementation to calculate the distance from the boundary 

### PR DESCRIPTION
**Pull request by pm-dimagi**
## Product Description
Written down the functions that calculate the distance from the boundary of the polygon 

## Technical Summary

Followed this document for the implementation of the code logic(https://docs.google.com/document/d/1mEqXmtuLSBnPGzcvxyEg-3X7BBytAlEFb8epkbevO64/edit?usp=sharing)

### Safety story
- Not able to do developer testing

### QA Plan
-Need proper testing to this function by creating the polygon and check the distance from that.
- Xpath function for the same is XPathBoundaryDistanceFunc

### Special deploy instructions

- [x] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [ ] This PR can be reverted after deploy with no further considerations.

### Review

- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### Duplicate PR
Automatically duplicate this PR as defined in [contributing.md](https://github.com/dimagi/commcare-core/blob/master/.github/contributing.md).

## Summary by CodeRabbit

- **New Features**
	- Added a new XPath function `boundaryDistance()` to calculate the distance to the closest polygon boundary
	- Introduced utility methods for geometric calculations involving polygons, including point-in-polygon checks and distance measurements

- **Technical Improvements**
	- Enhanced XPath expression capabilities with spatial analysis functionality
	- Added support for geometric computations in the core library
	
Release Note: Adds two new Xpath methods [closest-point-on-polygon](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143957589/CommCare+Functions#is-point-inside-polygon) and [is-point-inside-polygon](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143957589/CommCare+Functions#is-point-inside-polygon)

### CODE REVIEW

**Review by shubham1g5 — CHANGES_REQUESTED** [review]
I am not sure I understand how this function is supposed to use. Do you mind writing some unit-tests to demonstrate the function usage [in here](https://github.com/dimagi/commcare-core/blob/master/src/test/java/org/javarosa/xpath/test/XPathEvalTest.java#L170)

**Comment by ctsims** [comment]
Note: This will also need to be added here: https://github.com/dimagi/commcare-android/blob/0a85740209c3436e2ad61faedc1dc506ca5e5c7b/app/unit-tests/src/org/commcare/android/tests/processing/FormStorageTest.java#L363 for longitudinal testing to work

**Comment by ctsims** [comment]
Now that this is in a limited test, I wanted to make sure to write down a list of the things that I think need to be addressed in the structure before this is ready for production, since I might not be around or remember the context in full. 

- [ ] Implement appropriate sphere projection (great circle) math instead of planar math, and document the source of the approach
- [ ] Address edge cases for inputs (like a malformed polygon) to make sure they are communicated as effectively as possible, and add tests to clarify how invalid inputs caught
- [ ] Review the interfaces for the new functions (method signatures and names) - I don't think these are named super clearly right now, and I think the arguments should be in the other order (single input first, polygon second)
- [ ] Make a decision about the best set of data formats / models to be used for portable Geospatial geometries 
- [ ] Prepare the documentation for these functions, as well as any supporting documentation to use them (both things like the data format for the portable geometries and helpful usage documentation next to the reference docs that explain the use cases and how someone would adopt them. 
- [ ] Provide references from initial testing to confirm that these methods are useful and complete enough from a Product perspective, and worth adding to the generally available platform.

**Comment by pm-dimagi** [comment]
Hi Clayton,
I have noted this work will be picking up this week .
Thanks & Regards
Parth Mittal

**Comment by pm-dimagi** [comment]
> Now that this is in a limited test, I wanted to make sure to write down a list of the things that I think need to be addressed in the structure before this is ready for production, since I might not be around or remember the context in full.
> 
> * [ ]  Implement appropriate sphere projection (great circle) math instead of planar math, and document the source of the approach
I have used the library for this which is taking care of this library is https://github.com/locationtech/jts

> * [ ]  Address edge cases for inputs (like a malformed polygon) to make sure they are communicated as effectively as possible, and add tests to clarify how invalid inputs caught
I have created the test case for the following and asked chatgpt to create the test case to check the edge case 

> * [ ]  Review the interfaces for the new functions (method signatures and names) - I don't think these are named super clearly right now, and I think the arguments should be in the other order (single input first, polygon second)
made the necessary changes for the same 

> * [ ]  Make a decision about the best set of data formats / models to be used for portable Geospatial geometries
List of points will be in GeoJson format standard mapping polygon 

> * [ ]  Prepare the documentation for these functions, as well as any supporting documentation to use them (both things like the data format for the portable geometries and helpful usage documentation next to the reference docs that explain the use cases and how someone would adopt them.

> * [ ]  Provide references from initial testing to confirm that these methods are useful and complete enough from a Product perspective, and worth adding to the generally available platform.
Yes it is i have tested this on may test cases seprately and the outcome is from well used android library

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            Polygon polygon = PolygonUtils.createValidatedPolygon(Arrays.asList(coordinates));
+            // Casting and uncasting seems strange but is consistent with the codebase
+            GeoPointData pointData = new GeoPointData().cast(new UncastData(inputPoint));
+
+            return PolygonUtils.isPointInsideOrOnPolygon(pointData, polygon);
+        } catch (NumberFormatException e) {
+            throw new XPathTypeMismatchException("point-in-boundary() function requires arguments containing " +
+                    "numeric values only, but received arguments: " + inputPoint + " and " + inputPolygon);
+        } catch (IllegalArgumentException e) {
+            throw new XPathTypeMismatchException(e.getMessage());
```
Comment (shubham1g5) [inline src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java:56]: change to `XpathException`

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    protected Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        return isPointWithinBoundary(evaluatedArgs[0], evaluatedArgs[1]);
+    }
+
+    public static boolean isPointWithinBoundary(Object from, Object to) {
+        String inputPoint = (String)FunctionUtils.unpack(from);
+        String inputPolygon = (String)FunctionUtils.unpack(to);
+        if (inputPoint == null || "".equals(inputPoint) || inputPolygon == null || "".equals(inputPolygon)) {
+            return false;
```
Comment (shubham1g5) [inline src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java:43]: think we want the same behaviour here as `IllegalArgumentException` below as they essentially indicate an error with arguments.

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        try {
+            String[] coordinates = inputPolygon.split(" ");
+            Polygon polygon = PolygonUtils.createValidatedPolygon(Arrays.asList(coordinates));
+            // Casting and uncasting seems strange but is consistent with the codebase
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+public class XPathIsPointInsidePolygonFunc extends XPathFuncExpr {
+    public static final String NAME = "inside-polygon";
+    private static final int EXPECTED_ARG_COUNT = 2;
+
+    /**
+     * Returns true if the geopoint is inside the polygon, in meters, given objects to unpack.
+     * Ignores altitude and accuracy.
+     * Note that the arguments can be strings.
+     * Returns false if one of the arguments is null or the empty string.
+     */
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.utils.PolygonUtils;
+import org.javarosa.xpath.XPathTypeMismatchException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.Arrays;
+
+public class XPathIsPointInsidePolygonFunc extends XPathFuncExpr {
+    public static final String NAME = "inside-polygon";
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.javarosa.core.model.data.UncastData;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.utils.PolygonUtils;
+import org.javarosa.xpath.XPathTypeMismatchException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.Arrays;
+
+public class XPathIsPointInsidePolygonFunc extends XPathFuncExpr {
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathIsPointInsidePolygonFunc.java:41]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    @Override
+    protected Object evalBody(DataInstance model, EvaluationContext evalContext, Object[] evaluatedArgs) {
+        return closestPointToPolygon(evaluatedArgs[0], evaluatedArgs[1]);
+    }
+
+    public static String closestPointToPolygon(Object from, Object to) {
+        String inputPoint = (String)FunctionUtils.unpack(from);
+        String inputPolygon = (String)FunctionUtils.unpack(to);
+        if (inputPoint == null || "".equals(inputPoint) || inputPolygon == null || "".equals(inputPolygon)) {
+            return "";
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+            String[] coordinates = inputPolygon.split(" ");
+            Polygon polygon = PolygonUtils.createValidatedPolygon(Arrays.asList(coordinates));
+            // Casting and uncasting seems strange but is consistent with the codebase
+            GeoPointData pointData = new GeoPointData().cast(new UncastData(inputPoint));
+            return PolygonUtils.getClosestPointOnPolygon(pointData, polygon);
+        } catch (NumberFormatException e) {
+            throw new XPathTypeMismatchException("polygon-point() function requires arguments containing " +
+                    "numeric values only, but received arguments: " + inputPoint + " and " + inputPolygon);
+        } catch (IllegalArgumentException e) {
+            throw new XPathTypeMismatchException(e.getMessage());
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.utils.PolygonUtils;
+import org.javarosa.xpath.XPathTypeMismatchException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.Arrays;
+
+public class XPathClosestPointToPolygonFunc extends XPathFuncExpr {
+    public static final String NAME = "polygon-point";
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import org.javarosa.core.model.data.UncastData;
+import org.javarosa.core.model.instance.DataInstance;
+import org.javarosa.core.model.utils.PolygonUtils;
+import org.javarosa.xpath.XPathTypeMismatchException;
+import org.javarosa.xpath.parser.XPathSyntaxException;
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.Arrays;
+
+public class XPathClosestPointToPolygonFunc extends XPathFuncExpr {
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java:41]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/parser/ast/ASTNodeFunctionCall.java [hunk 11]
```diff
@@ -257,6 +259,10 @@ private static XPathFuncExpr buildFuncExpr(String name, XPathExpression[] args)
                 return new XPathDecryptStringFunc(args);
             case "json-property":
                 return new XPathJsonPropertyFunc(args);
+            case "polygon-point":
+                return new XPathClosestPointToPolygonFunc(args);
+            case "inside-polygon":
+                return new XPathIsPointInsidePolygonFunc(args);
```
[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/parser/ast/ASTNodeFunctionCall.java]: [a15919f](https://github.com/dimagi/commcare-core/pull/1455/commits/a15919fdf1d84ffd264e4d15e8a65a512178a194)

File: src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (shubham1g5) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java]: Class name needs to exactly reflect the method name here.

[the remark this answers is not in the document]
Comment (pm-dimagi) [reply src/main/java/org/javarosa/xpath/expr/XPathClosestPointToPolygonFunc.java]: [7d4a01b](https://github.com/dimagi/commcare-core/pull/1455/commits/7d4a01b67af6c7ef156a40a077f737fcb7eaf6b3)

**Review by shubham1g5 — COMMENTED** [review]
Looks good to be merged. But holding the approal here until we can get sign off from delivery on this. Also can you update the [wiki here](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143957589/CommCare+Functions) to reflect these new additions for CC 2.57 ?

**Review by shubham1g5 — APPROVED** [review]

**Comment by shubham1g5** [comment]
duplicate this PR ac492df619a8ae735d13c37c85ddda1a39d23e6d 7d4a01b67af6c7ef156a40a077f737fcb7eaf6b3

**Comment by shubham1g5** [comment]
duplicate this PR https://github.com/dimagi/commcare-core/commit/ac492df619a8ae735d13c37c85ddda1a39d23e6d https://github.com/dimagi/commcare-core/commit/7d4a01b67af6c7ef156a40a077f737fcb7eaf6b3
